File: apps/desktop/test/thread-follow-up-ui.test.ts
import { expect, test } from "bun:test";
import { i18n } from "../src/renderer/i18n";
import {
  canEscalateThreadFollowUp,
  formatThreadFollowUpPreview,
  isLiveFollowUpThreadStatus,
  mergeThreadFollowUp,
  queuedThreadFollowUps,
  shouldComposerUseFollowUpQueue,
} from "../src/renderer/thread-follow-up-ui";
import type { ThreadPendingFollowUp } from "../src/shared/ipc";
import {
  coreSupportsFollowUpEscalate,
  resolveFollowUpDeliveryModeForCore,
} from "../src/shared/thread-follow-up-core";

function followUp(id: string, patch: Partial<ThreadPendingFollowUp> = {}): ThreadPendingFollowUp {
  return {
    id,
    threadId: "thr_1",
    prompt: `message ${id}`,
    priority: "normal",
    status: "queued",
    deliveryMode: "queued",
    createdAt: `2024-01-01T00:00:0${id.length}.000Z`,
    updatedAt: "2024-01-01T00:00:00.000Z",
    ...patch,
  };
}

test("isLiveFollowUpThreadStatus only opens running and queued UI", () => {
  expect(isLiveFollowUpThreadStatus("running")).toBe(true);
  expect(isLiveFollowUpThreadStatus("queued")).toBe(true);
  expect(isLiveFollowUpThreadStatus("awaiting_plan")).toBe(false);
  expect(isLiveFollowUpThreadStatus("completed")).toBe(false);
});

test("shouldComposerUseFollowUpQueue while live, editing, or paused", () => {
  expect(shouldComposerUseFollowUpQueue({ status: "running" })).toBe(true);
  expect(shouldComposerUseFollowUpQueue({ status: "idle" })).toBe(false);
  expect(
    shouldComposerUseFollowUpQueue({
      status: "idle",
      followUpQueuePaused: true,
    }),
  ).toBe(true);
  expect(
    shouldComposerUseFollowUpQueue({
      status: "completed",
      editingFollowUpId: "fu_1",
    }),
  ).toBe(true);
});

test("queuedThreadFollowUps hides non-queued records and preserves stable priority order", () => {
  const normal = followUp("normal", { createdAt: "2024-01-01T00:00:01.000Z" });
  const cancelled = followUp("cancelled", { status: "cancelled" });
  const escalated = followUp("escalated", {
    priority: "escalated",
    deliveryMode: "interrupt_resume",
    createdAt: "2024-01-01T00:00:02.000Z",
  });

  expect(queuedThreadFollowUps([normal, cancelled, escalated]).map((item) => item.id)).toEqual([
    "escalated",
    "normal",
  ]);
});

test("mergeThreadFollowUp replaces existing records by id", () => {
  const original = followUp("same", { prompt: "旧消息" });
  const updated = followUp("same", { prompt: "已取消", status: "cancelled" });

  expect(mergeThreadFollowUp([original], updated)).toEqual([updated]);
});

test("formatThreadFollowUpPreview localizes image and empty defaults", async () => {
  await i18n.changeLanguage("en-US");
  const preview = formatThreadFollowUpPreview(
    followUp("with-image", {
      prompt: "a".repeat(140),
      attachments: [{ mediaType: "image/png", data: "abc" }],
    }),
  );

  expect(preview).toEndWith("... (1 image(s))");
  expect(preview.length).toBeLessThan(140);
  expect(
    formatThreadFollowUpPreview(
      followUp("images", {
        prompt: "",
        attachments: [{ mediaType: "image/png", data: "abc" }],
      }),
    ),
  ).toBe("1 image(s)");
  expect(formatThreadFollowUpPreview(followUp("empty", { prompt: "" }))).toBe("Empty follow-up message");
});

test("Guide stays clickable on an escalated row while the queue is paused", () => {
  expect(
    canEscalateThreadFollowUp({ priority: "normal", coreSupportsEscalate: true, queuePaused: true }),
  ).toBe(true);
  expect(
    canEscalateThreadFollowUp({
      priority: "escalated",
      coreSupportsEscalate: true,
      queuePaused: true,
    }),
  ).toBe(true);
  // Unpaused: an escalated row is already next in line, so the action is inert.
  expect(
    canEscalateThreadFollowUp({
      priority: "escalated",
      coreSupportsEscalate: true,
      queuePaused: false,
    }),
  ).toBe(false);
  expect(
    canEscalateThreadFollowUp({ priority: "normal", coreSupportsEscalate: false, queuePaused: true }),
  ).toBe(false);
});

test("follow-up UI allows escalate and preserves steer for mid-turn cores", () => {
  expect(coreSupportsFollowUpEscalate("acp")).toBe(true);
  expect(coreSupportsFollowUpEscalate("claude")).toBe(true);
  expect(coreSupportsFollowUpEscalate("pi")).toBe(true);
  expect(resolveFollowUpDeliveryModeForCore("acp", "steer")).toBe("steer");
  expect(resolveFollowUpDeliveryModeForCore("codex", "steer")).toBe("steer");
  expect(resolveFollowUpDeliveryModeForCore("pi", "steer")).toBe("steer");
});

File: apps/mobile/test/thread_follow_up_ui_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/widgets.dart';

import 'package:eco_mobile/core/models/thread_models.dart';
import 'package:eco_mobile/core/utils/thread_follow_up_ui.dart';
import 'package:eco_mobile/l10n/generated/app_localizations.dart';

final _zh = lookupAppLocalizations(const Locale('zh'));

ThreadPendingFollowUp followUp(
  String id, {
  String status = 'queued',
  String priority = 'normal',
  String createdAt = '2024-01-01T00:00:00.000Z',
  String prompt = 'message',
}) {
  return ThreadPendingFollowUp(
    id: id,
    threadId: 'thr_1',
    prompt: prompt,
    status: status,
    createdAt: createdAt,
    priority: priority,
  );
}

void main() {
  test('isFollowUpThreadLiveEvent matches follow-up topics and payloads', () {
    expect(
      isFollowUpThreadLiveEvent(
        kind: 'thread.follow_up',
        liveType: 'thread.follow_up.queued',
      ),
      isTrue,
    );
    expect(
      isFollowUpThreadLiveEvent(
        kind: 'thread.lifecycle',
        liveType: 'thread.follow_up.applied',
      ),
      isTrue,
    );
    expect(
      isFollowUpThreadLiveEvent(
        kind: 'thread.stream',
        liveType: 'thread.running',
        hasFollowUp: true,
      ),
      isTrue,
    );
    expect(
      isFollowUpThreadLiveEvent(
        kind: 'thread.stream',
        liveType: 'thread.running',
      ),
      isFalse,
    );
  });

  test('resolveThreadEventThreadId prefers envelope thread id', () {
    expect(
      resolveThreadEventThreadId(
        envelopeThreadId: 'thr_envelope',
        payloadThreadId: 'thr_payload',
      ),
      'thr_envelope',
    );
    expect(
      resolveThreadEventThreadId(payloadThreadId: 'thr_payload'),
      'thr_payload',
    );
    expect(resolveThreadEventThreadId(), isNull);
  });

  test('isLiveFollowUpThreadStatus only opens running and queued UI', () {
    expect(isLiveFollowUpThreadStatus('running'), isTrue);
    expect(isLiveFollowUpThreadStatus('queued'), isTrue);
    expect(isLiveFollowUpThreadStatus('awaiting_plan'), isFalse);
    expect(isLiveFollowUpThreadStatus('completed'), isFalse);
  });

  test('queuedThreadFollowUps hides non-queued records', () {
    final normal = followUp('normal');
    final cancelled = followUp('cancelled', status: 'cancelled');
    final escalated = followUp(
      'escalated',
      priority: 'escalated',
      createdAt: '2024-01-01T00:00:01.000Z',
    );

    expect(
      queuedThreadFollowUps([
        normal,
        cancelled,
        escalated,
      ]).map((item) => item.id),
      ['escalated', 'normal'],
    );
  });

  test('mergeThreadFollowUp replaces existing records by id', () {
    final original = followUp('same', prompt: '旧消息');
    final updated = followUp('same', prompt: '已取消', status: 'cancelled');

    expect(mergeThreadFollowUp([original], updated), [updated]);
  });

  test('formatThreadFollowUpPreview clips long prompts', () {
    final preview = formatThreadFollowUpPreview(
      followUp('long', prompt: 'a' * 130),
      _zh,
    );

    expect(preview.length, 120);
    expect(preview.endsWith('...'), isTrue);
  });

  test('formatThreadFollowUpPreview includes image count', () {
    final item = ThreadPendingFollowUp(
      id: 'image-follow-up',
      threadId: 'thread-1',
      prompt: '看一下',
      status: 'queued',
      createdAt: '2026-01-01T00:00:00.000Z',
      attachments: const [
        PromptImageAttachment(mediaType: 'image/jpeg', data: 'YWJj'),
      ],
    );

    expect(formatThreadFollowUpPreview(item, _zh), '看一下 (1 张图片)');
  });
}
